fdbclient/MutationLogReader.actor.h

```diff
@@ -0,0 +1,135 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#if defined(NO_INTELLISENSE) && !defined(FDBCLIENT_MUTATION_LOG_READER_ACTOR_G_H)
+#define FDBCLIENT_MUTATION_LOG_READER_ACTOR_G_H
+#include "fdbclient/MutationLogReader.actor.g.h"
+#elif !defined(FDBCLIENT_MUTATION_LOG_READER_ACTOR_H)
+#define FDBCLIENT_MUTATION_LOG_READER_ACTOR_H
+
+#include <deque>
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/NativeAPI.actor.h"
+#include "flow/flow.h"
+#include "flow/ActorCollection.h"
+#include "flow/actorcompiler.h" // has to be last include
+
+namespace mutation_log_reader {
+
+// RangeResultBlock is the wrapper of RangeResult. Each PipelinedReader maintains a Deque of RangeResultBlocks, from its
+// getRange results. MutationLogReader maintains a min heap of RangeResultBlocks, and provides RangeResult
+// partially in it to consumer.
+struct RangeResultBlock {
+	RangeResult result;
+	Version firstVersion; // version of first record, inclusive
+	Version lastVersion; // version of last record, inclusive
+	uint8_t hash; // points back to the PipelinedReader
+	int prefixLen; // size of keyspace, uid, and hash prefix
+	int indexToRead; // index of first unconsumed record
+
+	// When the consumer reads, provides (partial) RangeResult, from firstVersion to min(lastVersion, firstVersion
+	// rounded up to the nearest 1M), to ensure that no versions out of this RangeResultBlock can be in between.
+	Standalone<RangeResultRef> consume();
+
+	bool empty() { return indexToRead == result.size(); }
+
+	bool operator<(const RangeResultBlock& r) const {
+		// We want a min heap. The standard C++ priority queue is a max heap.
+		return firstVersion > r.firstVersion;
+	}
+};
+
+// PipelinedReader is the class actually doing range read (getRange). A MutationLogReader has 256 PipelinedReaders, each
+// in charge of one hash value from 0-255.
+class PipelinedReader {
+public:
+	PipelinedReader(uint8_t h, Version bv, Version ev, unsigned pd, Key p)
+	  : readerLimit(pd), hash(h), prefix(StringRef(&hash, sizeof(uint8_t)).withPrefix(p)), beginVersion(bv),
+	    endVersion(ev), currentBeginVersion(bv), pipelineDepth(pd) {}
+
+	void startReading(Database cx);
+	Future<Void> getNext(Database cx);
+	ACTOR static Future<Void> getNext_impl(PipelinedReader* self, Database cx);
+
+	void release() { readerLimit.release(); }
+
+	PromiseStream<RangeResultBlock> reads;
+	FlowLock readerLimit;
+	uint8_t hash;
+	Key prefix; // "\xff\x02/alog/UID/hash/" for restore, or "\xff\x02/blog/UID/hash/" for backup
+
+	Future<Void> done() { return reader; }
+
+private:
+	Version beginVersion, endVersion, currentBeginVersion;
+	unsigned pipelineDepth;
+	Future<Void> reader;
+};
+
+} // namespace mutation_log_reader
+
+// A MutationLogReader has 256 PipelinedReaders, each in charge of one hash value from 0-255. It keeps a min heap of
```

Comment:
The first sentence should explain what MLR does at a high level - provides a strictly version ordered stream of KV pairs that represent mutation log chunks written by the FDB backup log feature.